pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -978,58 +985,93 @@ func (a *Operator) updateNamespaceList(op *operatorsv1.OperatorGroup) ([]string,
 	return namespaceList, nil
 }
 
+func (a *Operator) clusterRoleNamePrefix(roleSuffix string) string {
+	return fmt.Sprintf("olm.operatorgroup.%s-", roleSuffix)
+}
+
 func (a *Operator) ensureOpGroupClusterRole(op *operatorsv1.OperatorGroup, suffix string, apis cache.APISet) error {
-	clusterRole := &rbacv1.ClusterRole{
+	// create target cluster role spec
+	var clusterRole *rbacv1.ClusterRole
+	clusterRoleNamePrefix := a.clusterRoleNamePrefix(suffix)
+	aggregationRule, err := a.getClusterRoleAggregationRule(apis, suffix)
+	if err != nil {
+		return err
+	}
+
+	clusterRole = &rbacv1.ClusterRole{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: strings.Join([]string{op.GetName(), suffix}, "-"),
+			Name: genName(clusterRoleNamePrefix),
 		},
+		AggregationRule: aggregationRule,
 	}
-	var selectors []metav1.LabelSelector
-	for api := range apis {
-		aggregationLabel, err := aggregationLabelFromAPIKey(api, suffix)
-		if err != nil {
-			return err
-		}
-		selectors = append(selectors, metav1.LabelSelector{
-			MatchLabels: map[string]string{
-				aggregationLabel: "true",
-			},
-		})
-	}
-	if len(selectors) > 0 {
-		clusterRole.AggregationRule = &rbacv1.AggregationRule{
-			ClusterRoleSelectors: selectors,
-		}
-	}
-	err := ownerutil.AddOwnerLabels(clusterRole, op)
-	if err != nil {
+
+	if err := ownerutil.AddOwnerLabels(clusterRole, op); err != nil {
 		return err
 	}
 
-	existingRole, err := a.lister.RbacV1().ClusterRoleLister().Get(clusterRole.Name)
-	if err != nil && !apierrors.IsNotFound(err) {
+	// get existing cluster role for this level (suffix: admin, edit, view))
+	existingRole, err := a.getExistingClusterRoleByOwnerAndLevel(op, suffix)
+	if err != nil {
```

Comment:
Do we need not found error handling here?

Reply:
We don't need to because underneath it's a list query. However, I've been thinking about this approach, and having a random string at the end _could_ lead to multiple ClusterRolesBeing created. Because we check against the existence of ClusterRoles owned by the operator group, we introduce a race condition. So, I've pivoted back to the strategy of using operator group information to generate the random string consistently. Originally, we were worried about using namespace/name beacuse it could lead to guessing attacks and leak sensitive cluster information (potentially). In this iteration, I'm using the resource's uid which should always be present (once the resource is created by the api server - which is when we care about it anyway) and is also unique and meaningless. This does affect testing because now we _need_ to define a UID.

@m1kola what do you think of the new approach?